### PR TITLE
Add more error handling plus some refactoring

### LINKED ISSUE #129: "no valid configuration file found (0 configs were invalid)"
Greetings! Testing out the platform, and enjoying things so far. Got this error in my Github Actions pipeline, and followed your request to post it for analysis. Maybe related to #112?

```
Run returntocorp/semgrep-action@v1
  with:
    publishToken: ***
    publishDeployment: 203
  env:
    GITHUB_TOKEN: ***
/usr/bin/docker run --name returntocorpsemgrepactionv1_351056 --label 179394 --workdir /github/workspace --rm -e GITHUB_TOKEN -e INPUT_PUBLISHTOKEN -e INPUT_PUBLISHDEPLOYMENT -e INPUT_CONFIG -e INPUT_GENERATESARIF -e HOME -e GITHUB_JOB -e GITHUB_REF -e GITHUB_SHA -e GITHUB_REPOSITORY -e GITHUB_REPOSITORY_OWNER -e GITHUB_RUN_ID -e GITHUB_RUN_NUMBER -e GITHUB_RETENTION_DAYS -e GITHUB_ACTOR -e GITHUB_WORKFLOW -e GITHUB_HEAD_REF -e GITHUB_BASE_REF -e GITHUB_EVENT_NAME -e GITHUB_SERVER_URL -e GITHUB_API_URL -e GITHUB_GRAPHQL_URL -e GITHUB_WORKSPACE -e GITHUB_ACTION -e GITHUB_EVENT_PATH -e GITHUB_ACTION_REPOSITORY -e GITHUB_ACTION_REF -e GITHUB_PATH -e GITHUB_ENV -e RUNNER_OS -e RUNNER_TOOL_CACHE -e RUNNER_TEMP -e RUNNER_WORKSPACE -e ACTIONS_RUNTIME_URL -e ACTIONS_RUNTIME_TOKEN -e ACTIONS_CACHE_URL -e GITHUB_ACTIONS=true -e CI=true -v "/var/run/docker.sock":"/var/run/docker.sock" -v "/home/runner/work/_temp/_github_home":"/github/home" -v "/home/runner/work/_temp/_github_workflow":"/github/workflow" -v "/home/runner/work/_temp/_runner_file_commands":"/github/file_commands" -v "/home/runner/work/semgrep-test-repo/semgrep-test-repo":"/github/workspace" returntocorp/semgrep-action:v1
=== detecting environment
| versions          - semgrep 0.32.0 on Python 3.7.9
| environment       - running in environment github-actions, triggering event is 'pull_request'
| manage            - logged in as deployment #203
=== setting up agent configuration
| policy            - using Getting Started
| using semgrep rules configured on the web UI
| using default path ignore rules of common test and dependency directories
| looking at 4 changed paths
| found 4 files in the paths to be scanned
=== looking for current issues in 4 files

=== failed command's STDOUT:

{"results": [], "errors": [{"type": "SemgrepError", "code": 7, "message": "no valid configuration file found (0 configs were invalid)"}]}


=== failed command's STDERR:

A new version of Semgrep is available. Please see https://github.com/returntocorp/semgrep#upgrading for more information.


Error: ROR] `/root/.local/bin/semgrep --skip-unknown-extensions --disable-nosem --json --no-rewrite-rule-ids --config /tmp/tmp3tby2xfe.yml more_fail.py other_feature.py .github/workflows/semgrep.yml should_fail.py` failed with exit code 7

This is an internal error, please file an issue at https://github.com/returntocorp/semgrep-action/issues/new/choose
and include any log output from above.
```

### PR DESCRIPTION
**Pull request by brendongo**
Fixes #129

### CODE REVIEW

File: src/semgrep_agent/utils.py [hunk 1]
```diff
@@ -28,10 +28,10 @@ def __init__(self, message: str) -> None:
 
 def debug_echo(text: str) -> None:
     """Print debug messages with context-specific debug formatting."""
-    if os.getenv("GITHUB_ACTIONS"):
-        prefix = "::debug::"
-    elif os.getenv("SEMGREP_AGENT_DEBUG"):
+    if os.getenv("SEMGREP_AGENT_DEBUG"):
```
Comment (brendongo) [inline src/semgrep_agent/utils.py:31]: Switched these two so if SEMGREP_AGENT_DEBUG is set to 1 in env variiable no need to also set the Github actions debug flag in secrets. Or is there a reason we want to put debug flag behind secrets permissioning @underyx ?

File: src/semgrep_agent/main.py [hunk 2]
```diff
@@ -80,11 +65,8 @@ def main(
     )
 
     # Get Metadata
-    Meta = detect_meta_environment()
-    meta_kwargs = {}
-    if baseline_ref:
-        meta_kwargs["cli_baseline_ref"] = baseline_ref
-    meta = Meta(config, **meta_kwargs)
+    meta = generate_meta_from_environment(baseline_ref)
```
Comment (nbrahms) [inline src/semgrep_agent/main.py:68]: Do we no longer allow baseline ref from the command line?

File: src/semgrep_agent/main.py [hunk 3]
```diff
@@ -95,18 +77,16 @@ def main(
...
         to_server = "" if publish_url == "https://semgrep.dev" else f" to {publish_url}"
         click.echo(
             get_aligned_command(
                 "manage", f"logged in{to_server} as deployment #{sapp.deployment_id}"
             ),
             err=True,
         )
-        policy_info = f"using {policy}" if policy else "unknown"
-        click.echo(get_aligned_command("policy", policy_info))
+        click.echo(get_aligned_command("policy", f"using {policy}"))
```
Comment (nbrahms) [inline src/semgrep_agent/main.py:89]: Is there always a policy now?

File: src/semgrep_agent/meta.py [hunk 4]
```diff
@@ -276,25 +275,19 @@ def pr_title(self) -> Optional[str]:
...
-
-    elif os.getenv("CI"):  # nosem
-        return GitMeta
+        return GitlabMeta()
 
     else:  # nosem
-        return GitMeta
+        if not baseline_ref:
+            click.echo(
+                "Note that no baseline-ref was passed and detected environment is not Github or Gitlab to will default to perfoming a full scan. If you want semgrep to only report on new findings from a pull request please set baseline-ref ",
```
Comment (nbrahms) [inline src/semgrep_agent/meta.py:290]: Not sure what you meant this to say?

File: src/semgrep_agent/semgrep_app.py [hunk 5]
```diff
@@ -105,6 +104,13 @@ def fetch_rules_text(self) -> str:
                 f"API server at {self.url} returned this error: {response.text}\n"
                 "Failed to get configured rules"
             )
+
+        # Can remove once server guarantees will always have at least one rule
+        parsed = yaml.load(response.text)
+        if not parsed["rules"]:
+            raise ActionFailure(
+                "No rules returned by server for this scan. Note that if a rule is not set to notify or block it is not returned by the server."
```
Comment (nbrahms) [inline src/semgrep_agent/semgrep_app.py:112]: :+1:

File: src/semgrep_agent/utils.py (hunk 1 above)
Comment (nbrahms) [reply src/semgrep_agent/utils.py:31]: I think GHA does some funky parsing

File: tests/test_semgrep_app.py [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+
+TEST_URL = "https://nonexist.semgrep.dev"
+DEPLOYMENT_ID = 0
+SCAN_ID = 5
+FAIL_MESSAGE = (
+    "Failed to send notifications for findings: no notification channels are enabled"
+)
+
+
+def mock_request_post(*args, **kwargs):
```
Comment (nbrahms) [inline tests/test_semgrep_app.py:14]: OMG thank you

**Review by nbrahms — APPROVED** [review]

File: src/semgrep_agent/meta.py (hunk 4 above)
Comment (chmccreery) [reply src/semgrep_agent/meta.py:290]: maybe "is not GitHub or GitLab so semgrep-action will default to performing a full scan"

**Comment by brendongo** [comment]
Looks like with merging tests are failing but the revert also fails tests. Looking.

File: src/semgrep_agent/main.py (hunk 2 above)
Comment (brendongo) [reply src/semgrep_agent/main.py:68]: There's no functionality change here afaict

File: src/semgrep_agent/main.py (hunk 3 above)
Comment (brendongo) [reply src/semgrep_agent/main.py:89]: Yeah changed the return to be non-optional

**Pull request merged**